content/en/blog/_posts/2021-11-18-persistentvolume-reclaim-policy-improvements.md

```diff
@@ -0,0 +1,192 @@
+```
+NAME                        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS               AGE
+example-vanilla-block-pvc   Bound    pvc-6791fdd4-5fad-438e-a7fb-16410363e3da   5Gi        RWO            example-vanilla-block-sc   19s
+```
+
+#### Delete PV
+
+When I try to delete a bound PV, the cluster blocks and the `kubectl` tool does
+not return back control to the shell; for example:
+
+```
+kubectl delete pv pvc-6791fdd4-5fad-438e-a7fb-16410363e3da
+```
+
+```
+persistentvolume "pvc-6791fdd4-5fad-438e-a7fb-16410363e3da" deleted
+^C
+```
+
+Retrieving the PV:
+```
+kubectl get pv pvc-6791fdd4-5fad-438e-a7fb-16410363e3da
+```
+
+It can be observed that the PV is in `Terminating` state
+```
+NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS        CLAIM                               STORAGECLASS               REASON   AGE
+pvc-6791fdd4-5fad-438e-a7fb-16410363e3da   5Gi        RWO            Delete           Terminating   default/example-vanilla-block-pvc   example-vanilla-block-sc            2m23s
+```
+
+#### Delete PVC
+
+```
+kubectl delete pvc example-vanilla-block-pvc
+```
+
+The following output is seen if the PVC gets successfully deleted:
+```
+persistentvolumeclaim "example-vanilla-block-pvc" deleted
+```
+
+The PV object from the cluster also gets deleted. When attempting to retrieve the PV
+it will be observed that the PV is no longer found:
+
+```
+kubectl get pv pvc-6791fdd4-5fad-438e-a7fb-16410363e3da
+```
+
+```
+Error from server (NotFound): persistentvolumes "pvc-6791fdd4-5fad-438e-a7fb-16410363e3da" not found
+```
+
+Although the PV is deleted the underlying storage resource is not deleted, and
+needs to be removed manually.
+
+To sum it up, the reclaim policy associated with the Persistent Volume is currently
+ignored under certain circumstance. For a `Bound` PV-PVC pair the ordering of PV-PVC
+deletion determines whether the PV reclaim policy is honored. The reclaim policy 
+is honored if the PVC is deleted first, however, if the PV is deleted prior to 
+deleting the PVC then the reclaim policy is not exercised. As a result of this behavior,
+the associated storage asset in the external infrastructure is not removed.
+
+## PV reclaim policy with Kubernetes v1.23
+
+The new behavior ensures that the underlying storage object is deleted from the backend when users attempt to delete a PV manually.
+
+**The fix is introduced as an alpha feature in 1.23 under the feature-gate `HonorPVReclaimPolicy`
+which is turned `off` by default**. You would also need to **update the `external-provisioner` 
+to the `v4.0.0` and enable the feature gate.**
+
+#### How does it work?
+
+The new behavior is achieved by adding a finalizer `external-provisioner.volume.kubernetes.io/finalizer` on new and existing PVs, the finalizer is only removed after the storage from backend is deleted.
```

Comment:
How does this work for in-tree persistent volume types? (perhaps the answer is: it doesn't and for in-tree storage handlers, there is no plan to fix the bug)

Reply:
The fix does not apply to in-tree volumes, there is an ongoing conversation to fix it for in-tree volumes in beta.